# Lab 2: From Raw Data to ML-Ready Data
## Data Cleaning & Exploratory Data Analysis — Telco Customer Churn

**Name:**   Amara Latif

**Roll Number:** 023-24-0012

**GitHub Repository:** _(link, added before submission)_

**Duration:** 3 Hours (independent, hands-on)

> This notebook is a **template**, not a tutorial. You already know Python, NumPy, Pandas, and Matplotlib from previous labs. Each section below states the objective and the questions you must answer — **you decide** which functions/plots are appropriate. Do not just run code without interpreting it: every visualization needs a written observation, and every cleaning decision needs a written justification.
>
> Fill in the empty code cells and the *Answer:* / *Observation:* placeholders directly in this notebook. Do not delete the Markdown headings — they are used for grading.

---

## Setup

Import the libraries you'll need and load the dataset. Use the Telco Customer Churn CSV provided for this lab.

In [7]:
# Import libraries
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

# Load the dataset
df = pd.read_csv("WA_Fn-UseC_-Telco-Customer-Churn.csv")   # update path/filename if different
df.head()


,customerID,gender,SeniorCitizen,Partner,Dependents,tenure,PhoneService,MultipleLines,InternetService,OnlineSecurity,...,DeviceProtection,TechSupport,StreamingTV,StreamingMovies,Contract,PaperlessBilling,PaymentMethod,MonthlyCharges,TotalCharges,Churn
0,7590-VHVEG,Female,0,Yes,No,1,No,No phone service,DSL,No,...,No,No,No,No,Month-to-month,Yes,Electronic check,29.85,29.85,No
1,5575-GNVDE,Male,0,No,No,34,Yes,No,DSL,Yes,...,Yes,No,No,No,One year,No,Mailed check,56.95,1889.5,No
2,3668-QPYBK,Male,0,No,No,2,Yes,No,DSL,Yes,...,No,No,No,No,Month-to-month,Yes,Mailed check,53.85,108.15,Yes
3,7795-CFOCW,Male,0,No,No,45,No,No phone service,DSL,Yes,...,Yes,Yes,No,No,One year,No,Bank transfer (automatic),42.30,1840.75,No
4,9237-HQITU,Female,0,No,No,2,Yes,No,Fiber optic,No,...,No,No,No,No,Month-to-month,Yes,Electronic check,70.70,151.65,Yes


---

## Part 1 — The Business Problem

A telecommunications company is losing customers to competitors — this is called **customer churn**. The company wants to understand which customers are likely to leave, so it can intervene early. Before any model can be built, the raw data must be understood, audited for quality issues, and explored.

**Tasks**
1. In 2–3 sentences, state the business problem in your own words, and propose what the target variable likely is.
2. List 5 columns you expect to be useful predictors of churn, and 2 columns you expect to be useless as predictors — justify each choice in one line.
3. Write down 2 questions you personally want this data to answer about customers who churn.

**Answer 1:**
The  telecommunications company is losing company to competitors and wants to understand which customers are likely to leave the company so that company can take action to reduce the customer churn. Each row in the dataset represent one customer and contain information about their services , contract and billing.The target variable is churn which shows whether the customer left or stayed.

**Answer 2:**
Useful
1.contract: it strongly affect cutomer churn.
2.Tenure: it shows how long a customer has been with the company.
3.monthlycharges:Higher monthly charges may be related to customer churn.
4.Internet service: Different types of internet services may have different churn rates.
5.payment method:Different payment methods can show different customer churn behavior.
Useless
1.customerID: it is only for customer identification and is not useful for predicting customer churn.
2.Gender:Has a weak relationship with churn and may not be useful for prediction.

**Answer 3:**
Two questions can be:
Q1: Which type of contract has highest rate of customer churn rate?
Q2: Do customers with shorter tenure churn more frequently than longer tenure?

---

## Part 2 — Exploring the Dataset Structure

Before cleaning or analyzing anything, get an accurate picture of the dataset's shape and contents. Running a command is not the same as understanding its output — you must interpret what you see.

*Concepts/functions you may find useful:* `df.shape`, `df.columns`, `df.dtypes`, `df.info()`, `df.head()`, `df.sample(n)`, `df[col].unique()`, `df[col].nunique()`, `df.describe(include='all')`

**Tasks**
4. How many customers and how many features are in this dataset? List the numerical features and the categorical features separately.
5. Identify any column(s) that are simply identifiers rather than predictive features, and any column(s) whose data type looks wrong for what it represents.
6. Pick 3 categorical columns and report their unique values — do any contain unexpected or inconsistent categories?

In [9]:
# Task 4 — dataset shape, column names, data types
print("Shape:" ,df.shape)
print("Columns:" ,df.columns.tolist())
print("Data types:" ,df.dtypes)

#numerical features
# SeniorCitizen
# tenure
# MonthlyCharges

#Categorical features
# gender
# Partner
# Dependents
# PhoneService
# MultipleLines
# InternetService
# OnlineSecurity
# OnlineBackup
# DeviceProtection
# TechSupport
# StreamingTV
# StreamingMovies
# Contract
# PaperlessBilling
# PaymentMethod
# Churn

Shape: (7043, 21)
Columns: ['customerID', 'gender', 'SeniorCitizen', 'Partner', 'Dependents', 'tenure', 'PhoneService', 'MultipleLines', 'InternetService', 'OnlineSecurity', 'OnlineBackup', 'DeviceProtection', 'TechSupport', 'StreamingTV', 'StreamingMovies', 'Contract', 'PaperlessBilling', 'PaymentMethod', 'MonthlyCharges', 'TotalCharges', 'Churn']
Data types: customerID              str
gender                  str
SeniorCitizen         int64
Partner                 str
Dependents              str
tenure                int64
PhoneService            str
MultipleLines           str
InternetService         str
OnlineSecurity          str
OnlineBackup            str
DeviceProtection        str
TechSupport             str
StreamingTV             str
StreamingMovies         str
Contract                str
PaperlessBilling        str
PaymentMethod           str
MonthlyCharges      float64
TotalCharges            str
Churn                   str
dtype: object


In [ ]:
# Task 5 — identify identifier columns / wrong data types
print(df.dtypes)

#TotalCharges has incorrect datatype. it is currently stored as object but it represents numerical charges,
# so it should be converted to float(numeric) datatype.

customerID              str
gender                  str
SeniorCitizen         int64
Partner                 str
Dependents              str
tenure                int64
PhoneService            str
MultipleLines           str
InternetService         str
OnlineSecurity          str
OnlineBackup            str
DeviceProtection        str
TechSupport             str
StreamingTV             str
StreamingMovies         str
Contract                str
PaperlessBilling        str
PaymentMethod           str
MonthlyCharges      float64
TotalCharges            str
Churn                   str
dtype: object


In [14]:
# Task 6 — unique values in 3 categorical columns
print("Contract")
print(df["Contract"].unique())

print("internet service")
print(df["InternetService"].unique())

print("PaymentMethod:")
print(df["PaymentMethod"].unique())

Contract
<ArrowStringArray>
['Month-to-month', 'One year', 'Two year']
Length: 3, dtype: str
internet service
<ArrowStringArray>
['DSL', 'Fiber optic', 'No']
Length: 3, dtype: str
PaymentMethod:
<ArrowStringArray>
[         'Electronic check',              'Mailed check',
 'Bank transfer (automatic)',   'Credit card (automatic)']
Length: 4, dtype: str


**Observations (Part 2):**
<!-- The dataset contains 7043 rows and 21 columns. The numerical features include SeniorCitizen, tenure, and MonthlyCharges, while most of the remaining columns are categorical. TotalCharges is stored as an object/string even though it represents numerical charges, so its data type needs to be changed. customerID is an identifier and does not provide meaningful predictive information. The categorical features such as Contract, InternetService, and PaymentMethod have consistent categories with no obvious spelling variations. -->

---

## Part 3 — Investigating Data Quality

A dataset can look clean at a glance and still hide serious problems. You must actively search for missing values, duplicates, and suspicious values rather than assume there are none.

*Concepts/functions you may find useful:* `df.isnull().sum()`, `df.isnull().mean()*100`, `df.duplicated().sum()`, `df['id_col'].duplicated().sum()`, `df[col].value_counts()`

**Tasks**
7. Report missing values as both a count and a percentage per column. Which missing-value problems do you think require action, and which don't? Justify your reasoning.
8. Check for duplicate rows and duplicate customer IDs separately — are they the same issue or different issues?
9. Find at least one column with a suspicious, invalid, or inconsistently formatted value, and describe how you found it.

In [20]:
# Task 7 — missing values: count and percentage per column
missing_count = df.isnull().sum()

missing_percentage = df.isnull().mean() * 100

missing = pd.DataFrame({
    "Missing count": missing_count ,
    "Missing percentage ":missing_percentage
})
print(missing)
print(df["TotalCharges"].str.strip().eq("").sum())

                  Missing count  Missing percentage 
customerID                    0                  0.0
gender                        0                  0.0
SeniorCitizen                 0                  0.0
Partner                       0                  0.0
Dependents                    0                  0.0
tenure                        0                  0.0
PhoneService                  0                  0.0
MultipleLines                 0                  0.0
InternetService               0                  0.0
OnlineSecurity                0                  0.0
OnlineBackup                  0                  0.0
DeviceProtection              0                  0.0
TechSupport                   0                  0.0
StreamingTV                   0                  0.0
StreamingMovies               0                  0.0
Contract                      0                  0.0
PaperlessBilling              0                  0.0
PaymentMethod                 0               

In [27]:
# Task 8 — duplicate rows vs duplicate IDs
print("Duplicated rows",df.duplicated().sum())

print("duplicated customer ID's:" ,df["customerID"].duplicated().sum())

Duplicated rows 0
duplicated customer ID's: 0


In [29]:
# Task 9 — investigate a suspicious / invalid value

print(df[df["TotalCharges"].str.strip()== ""][[
    "customerID",
    "tenure",
    "MonthlyCharges",
    "TotalCharges"
]] )

      customerID  tenure  MonthlyCharges TotalCharges
488   4472-LVYGI       0           52.55             
753   3115-CZMZD       0           20.25             
936   5709-LVOEQ       0           80.85             
1082  4367-NUYAO       0           25.75             
1340  1371-DWPAZ       0           56.05             
3331  7644-OMVMY       0           19.85             
3826  3213-VVOLG       0           25.35             
4380  2520-SGTTA       0           20.00             
5218  2923-ARZLG       0           19.70             
6670  4075-WKNIU       0           73.35             
6754  2775-SEFEE       0           61.90             


**Answer 7 (which missing-value problems need action, and why):**
TotalCharges has missing values that need action because it is a numerical feature needed for analysis and prediction. The blank values should be converted to missing values and then handled appropriately.

**Answer 8 (duplicate rows vs duplicate IDs — same issue or different?):**
Duplicate rows and duplicate IDs are different issues. Duplicate rows mean the entire record is repeated, while duplicate IDs mean the same customer ID appears more than once. A customer can have multiple records without the whole row being duplicated.
    
**Answer 9 (suspicious value found and how):**
A suspicious value was found in TotalCharges: some values were blank even though the column is supposed to contain numerical charges. This was identified by checking the data type and missing/blank values.


---

## Part 4 — Data Cleaning Decisions

Cleaning is a series of justified decisions, not a fixed recipe. For every issue you fix, you must be able to explain why you chose that method over an alternative.

*Concepts/functions you may find useful:* `pd.to_numeric(s, errors='coerce')`, `df.fillna(value)`, `df.dropna()`, `df.drop_duplicates()`, `df[col].str.strip()`, `df.astype(dtype)`

**Tasks**
10. For every data-quality issue you found in Part 3, record: Problem → Decision → Method Used → Reason.
11. Apply your cleaning decisions to produce an intermediate cleaned dataframe, and confirm (with code output) that each issue is actually resolved.

**Task 10 — Cleaning Decision Log**

| Problem              | Decision           | Method Used            | Reason |
|TotalCharges is object|Convert to numeric  |pd.to_numeric()         |It represents money|
|11 blank TotalCharges |Convert to 0        |isfillna(0)             |These customers have tenure 0 |
|Duplicate rows        |No action needed    |Check only              |There are 0 duplicates|
|customerID            |Remove              |drop                    | Not useful predictive features|

In [30]:
# Task 11 — apply your cleaning decisions here
clean_df = df.copy()

# Remove extra spaces from TotalCharges
clean_df["TotalCharges"] = clean_df["TotalCharges"].str.strip()

# Convert TotalCharges to numeric
clean_df["TotalCharges"] = pd.to_numeric(
    clean_df["TotalCharges"],
    errors="coerce"
)

# Fill missing TotalCharges with 0
clean_df["TotalCharges"] = clean_df["TotalCharges"].fillna(0)

# Remove customerID
clean_df = clean_df.drop(columns=["customerID"])

print(clean_df.head())

   gender  SeniorCitizen Partner Dependents  tenure PhoneService  \
0  Female              0     Yes         No       1           No   
1    Male              0      No         No      34          Yes   
2    Male              0      No         No       2          Yes   
3    Male              0      No         No      45           No   
4  Female              0      No         No       2          Yes   

      MultipleLines InternetService OnlineSecurity OnlineBackup  \
0  No phone service             DSL             No          Yes   
1                No             DSL            Yes           No   
2                No             DSL            Yes          Yes   
3  No phone service             DSL            Yes           No   
4                No     Fiber optic             No           No   

  DeviceProtection TechSupport StreamingTV StreamingMovies        Contract  \
0               No          No          No              No  Month-to-month   
1              Yes          No  

In [31]:
# Task 11 (continued) — confirm each issue is resolved (e.g. re-check isnull().sum(), duplicated().sum(), dtypes)
print("Missing values")
print(clean_df.isnull().sum())

print("Duplicate rows")
print(clean_df.duplicated().sum())

print("Datatypes")
print(clean_df.dtypes)

Missing values
gender              0
SeniorCitizen       0
Partner             0
Dependents          0
tenure              0
PhoneService        0
MultipleLines       0
InternetService     0
OnlineSecurity      0
OnlineBackup        0
DeviceProtection    0
TechSupport         0
StreamingTV         0
StreamingMovies     0
Contract            0
PaperlessBilling    0
PaymentMethod       0
MonthlyCharges      0
TotalCharges        0
Churn               0
dtype: int64
Duplicate rows
22
Datatypes
gender                  str
SeniorCitizen         int64
Partner                 str
Dependents              str
tenure                int64
PhoneService            str
MultipleLines           str
InternetService         str
OnlineSecurity          str
OnlineBackup            str
DeviceProtection        str
TechSupport             str
StreamingTV             str
StreamingMovies         str
Contract                str
PaperlessBilling        str
PaymentMethod           str
MonthlyCharges      float64


---

## Part 5 — Univariate EDA

Look at individual variables in isolation before looking at relationships. Every chart must be followed by a written interpretation — a graph with no explanation earns no credit.

*Concepts/functions you may find useful:* `df[col].plot(kind='hist')` / `plt.hist()`, `df.boxplot(column=col)`, `df[col].value_counts().plot(kind='bar')`

**Tasks**
12. Visualize the distribution of at least 2 numerical features and 2 categorical features. For each chart, write 1–2 sentences on what you learned.
13. Do any numerical features show outliers or a skewed distribution? What might explain this?

In [ ]:
# Task 12 — numerical feature 1: distribution plot
plt.figure(figsize=(8,5))

plt.hist(clean_df("Tenure"), bins=20)

plt.ylabel("Tenure(Months)")
plt.xlabel("Number of customers ")
plt.title("Distribution of customer tenure")
plt.show()

**Observation:** customers are distributed across different tenure periods.Substantial number of relatively new customers. The distribution is slightly right_skewed.

In [ ]:
# Task 12 — numerical feature 2: distribution plot
plt.figure(figsize =(8,5))
plt.hist(clean_df["MonthlyCharges"] , bins=20)

plt.xlabel("Monthly charges")
plt.ylabel("number of customers")
plt.title("Distribution of monthly charges")

plt.show()

**Observation:** Monthly charges range from about 18.25 to 118.75. The distribution is fairly spread out and does not show extreme outliers.

In [ ]:
# Task 12 — categorical feature 1: distribution plot
clean_df["Contract"].values_count().plot(
    kind ="bar",
    figsize=(8,5)
)

plt.xlabel("Contract")
plt.ylabel("Number of Customers")
plt.title("Customer Contract Distribution")

plt.show()

**Observation:** Month-to-month contracts are the most common, while two-year contracts are less common.

In [ ]:
# Task 12 — categorical feature 2: distribution plot
clean_df["InternetService"].value_counts().plot(
    kind="bar",
    figsize=(8,5)
)

plt.xlabel("Internet Service")
plt.ylabel("Number of Customers")
plt.title("Internet Service Distribution")

plt.show()

**Observation:** Fiber optic is the most common internet service, followed by DSL. A smaller group does not use internet service.

**Answer 13 (outliers / skew and possible explanation):**
There are no major extreme outliers. TotalCharges is somewhat right-skewed because customers with longer tenure accumulate higher total charges.

---

## Part 6 — Target Variable Analysis

The `Churn` column is what you will eventually predict. Before Lab 3, you need to understand its distribution — this has direct consequences for how a model should later be trained and evaluated.

*Concepts/functions you may find useful:* `df['Churn'].value_counts()`, `df['Churn'].value_counts(normalize=True)`

**Tasks**
14. Report the number and percentage of churned vs. non-churned customers. Is the target balanced or imbalanced?
15. In 2–3 sentences, explain why class imbalance might matter when a classification model is trained later — do not describe how to fix it, just why it matters.

In [8]:
clean_df = df.copy()

clean_df["TotalCharges"] = clean_df["TotalCharges"].str.strip()

clean_df["TotalCharges"] = pd.to_numeric(
    clean_df["TotalCharges"],
    errors="coerce"
)

clean_df["TotalCharges"] = clean_df["TotalCharges"].fillna(0)

clean_df = clean_df.drop(columns=["customerID"])

In [9]:
# Task 14 — churn counts and percentages

churn_counts = clean_df["Churn"].value_counts()

churn_percent = clean_df["Churn"].value_counts(
    normalize=True
) * 100

print("Counts:")
print(churn_counts)

print("\nPercentages:")
print(churn_percent)

Counts:
Churn
No     5174
Yes    1869
Name: count, dtype: int64

Percentages:
Churn
No     73.463013
Yes    26.536987
Name: proportion, dtype: float64


**Answer 14 (balanced or imbalanced?):**
The target variable is imbalanced because approximately 73.46% of customers did not churn while only 26.54% churned.
**Answer 15 (why imbalance might matter later):**
Class imbalance matters because a model may perform well on the majority class while performing poorly on the minority class. In this dataset, a model that mostly predicts "No Churn" could appear accurate but fail to identify many customers who actually churn.

---

## Part 7 — Bivariate Analysis: Features vs. Churn

Now connect individual features to the target. You choose the appropriate visualization or summary for each feature type — a categorical feature is not explored the same way as a numerical one.

*Concepts/functions you may find useful:* `pd.crosstab(df[col], df['Churn'], normalize='index')`, `df.groupby(col)['Churn']...`, `sns.boxplot(x='Churn', y=col, data=df)`

**Tasks**
16. Investigate at least 2 categorical features (e.g. Contract, InternetService, PaymentMethod) against Churn. For each, record: Observation / Evidence / Possible Explanation.
17. Investigate at least 2 numerical features (e.g. tenure, MonthlyCharges) against Churn using the same Observation / Evidence / Possible Explanation format.

In [ ]:
# Task 16 — categorical feature 1 vs Churn
contract_churn = pd.crosstab(
    clean_df["Contract"],
    clean_df["Churn"],
    normalize="index"
) * 100

print(contract_churn)

**Observation:** Month-to-month customers have a much higher churn rate than customers with one-year or two-year contracts.

**Evidence:** Month-to-month churn = 42.71%
Two-year churn = 2.83%

**Possible Explanation:**Longer contracts may provide greater commitment and stability, while month-to-month customers can leave more easily.

In [ ]:
# Task 16 — categorical feature 2 vs Churn
internet_churn = pd.crosstab(
    clean_df["InternetService"],
    clean_df["Churn"],
    normalize="index"
) * 100

print(internet_churn)

**Observation:** Fiber-optic customers have the highest churn rate.
**Evidence:** Fiber optic churn ≈ 41.89%
**Possible Explanation:**Fiber-optic customers may have higher monthly charges or different service expectations.

In [ ]:
# Task 17 — numerical feature 1 vs Churn
print(
    clean_df.groupby("Churn")["tenure"].mean()
)

**Observation:** Customers who churn tend to have much shorter tenure.
**Evidence:**Non-churned average tenure ≈ 37.57 months

Churned average tenure ≈ 17.98 months
**Possible Explanation:**Newer customers may not yet have strong loyalty or long-term commitment.

In [ ]:
# Task 17 — numerical feature 2 vs Churn
print(
    clean_df.groupby("Churn")["MonthlyCharges"].mean()
)

**Observation:**Customers who churn have higher average monthly charges.
**Evidence:**Churned customers:

Average MonthlyCharges ≈ 74.44

Non-churned:≈ 61.27
**Possible Explanation:** Higher monthly costs may contribute to dissatisfaction or customers switching to competitors.

---

## Part 8 — Correlation & Feature Relationships

Numerical features can be related to each other, not just to the target. Strongly correlated features can cause redundancy problems later, so it's worth knowing about them now.

*Concepts/functions you may find useful:* `df.corr(numeric_only=True)`, `sns.heatmap(corr_matrix, annot=True)`

**Tasks**
18. Compute and visualize the correlation matrix of the numerical features. Which pairs appear most correlated?
19. Does correlation imply causation here? Could any highly correlated features cause problems for a future model? Answer in 2–3 sentences.

In [13]:
# Task 18 — correlation matrix and heatmap
corr_df= clean_df.copy()
corr_matrix = corr_df.corr(numeric_only=True)
print(corr_matrix)

                SeniorCitizen    tenure  MonthlyCharges  TotalCharges
SeniorCitizen        1.000000  0.016567        0.220173      0.103006
tenure               0.016567  1.000000        0.247900      0.826178
MonthlyCharges       0.220173  0.247900        1.000000      0.651174
TotalCharges         0.103006  0.826178        0.651174      1.000000


**Answer 18 (most correlated pairs):**
tenure ↔ TotalCharges
≈ 0.826

MonthlyCharges ↔ TotalCharges
≈ 0.651
**Answer 19 (correlation vs. causation; potential problems):**
Correlation does not mean that one variable directly causes another. For example, tenure and TotalCharges are strongly correlated because customers who stay longer generally accumulate more charges. Highly correlated features can also create redundancy in a future ML model.

---

## Part 9 — Data Leakage Awareness

In Lab 3 you will train a classifier. Before that, you must check whether any column contains information that would not realistically be available at the moment of prediction — using such a column would make a model look artificially accurate.

**Tasks**
20. Inspect the columns for anything that looks like it could only be known after a customer has already churned, or that is purely an identifier. List any suspects and explain your reasoning.
21. In 2–3 sentences, explain what would go wrong if such a column were left in the model's input features.

**Answer 20 (suspect columns and reasoning):**
customerID should be removed wbecause it is only an identifier and does not provide any meaningful predictive information, Churn should also not be  used as an input feature because it is the target variable that the model is supposed to predict.

**Answer 21 (what would go wrong):**
Data leakage can make a model appear much more accurate during training and evaluation that it really is. The model may use information that would not actually be available when maiing a real prediction , resulting in unrealistic performance.


---

## Part 10 — ML-Readiness Checklist

Summarize every column's fate in a single table. This is the bridge between EDA and modeling — Lab 3 will use this table directly.

**Task**
22. Complete the ML-readiness table for **every** column in the dataset.

| Column           | Type        | Role    | Action | Reason                         |
| ---------------- | ----------- | ------- | ------ | ------------------------------ |
| customerID       | Identifier  | —       | Remove | Only identifies customer       |
| gender           | Categorical | Feature | Keep   | Customer characteristic        |
| SeniorCitizen    | Numerical   | Feature | Keep   | Customer characteristic        |
| Partner          | Categorical | Feature | Keep   | Customer characteristic        |
| Dependents       | Categorical | Feature | Keep   | Customer characteristic        |
| tenure           | Numerical   | Feature | Keep   | Strong relationship with churn |
| PhoneService     | Categorical | Feature | Keep   | Service information            |
| MultipleLines    | Categorical | Feature | Keep   | Service information            |
| InternetService  | Categorical | Feature | Keep   | Strong churn differences       |
| OnlineSecurity   | Categorical | Feature | Keep   | Service information            |
| OnlineBackup     | Categorical | Feature | Keep   | Service information            |
| DeviceProtection | Categorical | Feature | Keep   | Service information            |
| TechSupport      | Categorical | Feature | Keep   | Service information            |
| StreamingTV      | Categorical | Feature | Keep   | Service information            |
| StreamingMovies  | Categorical | Feature | Keep   | Service information            |
| Contract         | Categorical | Feature | Keep   | Strong churn relationship      |
| PaperlessBilling | Categorical | Feature | Keep   | Billing information            |
| PaymentMethod    | Categorical | Feature | Keep   | Churn differences              |
| MonthlyCharges   | Numerical   | Feature | Keep   | Billing information            |
| TotalCharges     | Numerical   | Feature | Keep   | Billing information            |
| Churn            | Categorical | Target  | Keep   | Variable to predict            |



_(Add one row per remaining column — copy the pattern above.)_

---

## Part 11 — Building the Clean Dataset

Bring together every cleaning decision into one final dataframe, `clean_df`, that Lab 3 can load directly.

> Do **not** split into train/test, encode categorical variables, or train any model — that happens in Lab 3.

**Tasks**
23. Produce `clean_df` with correct data types, missing values handled, duplicates handled, and irrelevant identifier columns removed where justified. Save it (e.g. `clean_churn.csv`) alongside your notebook.
24. Print `df.info()` (or equivalent) for `clean_df` as final proof that every issue identified earlier has been resolved.

In [18]:
# Task 23 — assemble clean_df from your cleaning decisions
clean_df = df.copy()

# Convert to numeric
clean_df["TotalCharges"] = pd.to_numeric(
    clean_df["TotalCharges"] ,
    errors="coerce"
)
# Fill missing TotalCharges
clean_df["TotalCharges"] = clean_df["TotalCharges"].fillna(0)

# Remove identifier
clean_df = clean_df.drop(columns = ["customerID"])

# Save the cleaned dataset
clean_df.to_csv("clean_churn.csv", index=False)
print(clean_df.shape)


(7043, 20)


In [19]:
# Task 24 — final proof: info() / isnull().sum() / duplicated().sum() on clean_df
print(clean_df.info())

print("\nMissing values:")
print(clean_df.isnull().sum())

print("\nDuplicate rows:")
print(clean_df.duplicated().sum())

<class 'pandas.DataFrame'>
RangeIndex: 7043 entries, 0 to 7042
Data columns (total 20 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   gender            7043 non-null   str    
 1   SeniorCitizen     7043 non-null   int64  
 2   Partner           7043 non-null   str    
 3   Dependents        7043 non-null   str    
 4   tenure            7043 non-null   int64  
 5   PhoneService      7043 non-null   str    
 6   MultipleLines     7043 non-null   str    
 7   InternetService   7043 non-null   str    
 8   OnlineSecurity    7043 non-null   str    
 9   OnlineBackup      7043 non-null   str    
 10  DeviceProtection  7043 non-null   str    
 11  TechSupport       7043 non-null   str    
 12  StreamingTV       7043 non-null   str    
 13  StreamingMovies   7043 non-null   str    
 14  Contract          7043 non-null   str    
 15  PaperlessBilling  7043 non-null   str    
 16  PaymentMethod     7043 non-null   str    
 17  Monthl

---

## Part 12 — Key Findings & Final Reflection

Step back from the code and summarize what you actually learned about this dataset and this problem.

**Task 25 — Five Important Findings**

Repeat this block for each of your five findings.

**Finding 1**
- Observation: Month-to-month customers churn much more frequently.
- Evidence: month-to-month churn rate ≈ 42.71%, compared with 2.83% for two-year contracts.
- Interpretation: _..._

**Finding 2**
- Observation:  Churned customers have shorter tenure.

- Evidence:Average tenure is about 17.98 months for churned customers versus 37.57 months for non-churned customers.

- Interpretation:Newer customers appear more likely to leave.

**Finding 3**
- Observation: Fiber-optic customers have high churn.
- Evidence: Fiber-optic churn rate ≈ 41.89%.
- Interpretation: Internet service type may be an important predictor of churn.

**Finding 4**
- Observation: Churned customers have higher monthly charges.

- Evidence: Average monthly charge is about 74.44 for churned customers versus 61.27 for non-churned customers.

- Interpretation: Higher monthly cost may be associated with customer churn.

**Finding 5**
- Observation: TotalCharges is strongly related to tenure.
- Evidence: Correlation ≈ 0.826.
- Interpretation: Customers with longer tenure naturally accumulate higher total charges.

---

**Task 26 — Final Reflection Questions**

1. What was the most important data-quality issue you found?
   Total charges are stored as an object instead of numerical column, and 11 customers had blank values.These values needed to be converted to numeric and handled appropriately.

2. What was the most interesting pattern you discovered?
  Contract type shows strong difference in churn. Month-to-month customers had a much higher churn rate than customers with longer contracts.

3. Which feature seems most useful for predicting churn?
   contract appear to be one of the most useful  features because its categories have very useful churn rates

4. Which feature should probably not be used, and why?
   customerID should not be used because it is only an identifier and does not provide meaningful predictive information.

5. What preprocessing will be required before ML modeling (conceptually — do not implement it)?
   Categorical variables will need to be encoded , numerical features may need scaling depending on ML algorithm and the target varibale need to be converted into as suuitable format.

6. What would you do next if you had another 3 hours?
   I would investigate more feature combinations, study churn across different customer segments, examine additional visualizations, and compare the strongest predictors before starting ML modeling.
  

---

## Bonus Tasks (Optional)

- Create one additional meaningful visualization not required above.
- Investigate a surprising customer subgroup you didn't expect to find.
- Calculate churn rate across a combination of 2+ customer segments at once.
- Identify one possible feature-engineering opportunity for Lab 3.
- Identify one possible source of bias in this dataset.

In [ ]:
# Bonus — optional exploration

1.A heatmap of churn rate by Contract and InternetService is useful because it shows the effect of two variables together.
# Additional Visualization: Churn Rate by Contract and Internet Service

pivot = pd.crosstab(
    clean_intermediate["Contract"],
    clean_intermediate["InternetService"],
    values=(clean_intermediate["Churn"] == "Yes"),
    aggfunc="mean"
) * 100

plt.figure(figsize=(8, 5))
sns.heatmap(pivot, annot=True, fmt=".1f", cmap="Blues")
plt.title("Churn Rate by Contract and Internet Service")
plt.xlabel("Internet Service")
plt.ylabel("Contract")
plt.show()

Interpretation: The combination of month-to-month contract + fiber optic internet has the highest churn rate,
making this subgroup particularly important.

2.A surprising subgroup is customers with no partner and no dependents.
Their churn rate is about 34.24%, compared with only 14.24% for customers who have both a partner and dependents.
This suggests that customers without family-related connections may be more likely to leave. This is an association, not proof that family status causes churn.

3.| Contract       | Internet Service | Churn Rate |
| -------------- | ---------------- | ---------: |
| Month-to-month | Fiber optic      | **54.61%** |
| Month-to-month | DSL              | **32.22%** |
| One year       | Fiber optic      | **19.29%** |
| One year       | DSL              |  **9.30%** |
| Two year       | Fiber optic      |  **7.23%** |
| Two year       | DSL              |  **1.91%** |
Finding: The highest-risk group is month-to-month customers with fiber optic service (54.61%). 
This is much higher than the overall churn rate of about 26.54%.

4.df["TenureGroup"] = pd.cut(
    df["tenure"],
    bins=[-1, 12, 24, 48, 72],
    labels=["0-12 months", "13-24 months", "25-48 months", "49-72 months"]
)

This converts the numerical tenure variable into meaningful customer-life-stage groups. 
Since churned customers have a much shorter average tenure (17.98 months vs. 37.57 months), this feature could help a churn prediction model.

5.One possible source of bias is the dataset comes from a single telecommunications company, 
so its customer behavior may not represent customers from other companies or markets.
 like, the high churn among fiber-optic customers (41.89%) may be specific to this company's pricing, 
service quality, or customer base rather than a general rule about fiber customers.


---

## Submission Checklist

Before submitting, confirm you have:

- [ ] Completed all 26 tasks above, with written interpretation for every chart and decision
- [ ] A working `clean_df`, saved as `clean_churn.csv`
- [ ] A completed ML-readiness table (Part 10) covering every column
- [ ] Five Key Findings written (Part 12)
- [ ] Answered all six Final Reflection questions
- [ ] Committed and pushed this notebook + `clean_churn.csv` to your GitHub repository
- [ ] Added your GitHub repository link at the top of this notebook